### ISSUE TITLE
in method Range#includes?(value), when value can't compare range item and in case statement, union type

### ISSUE THREAD

**Issue by nemorami** [crystal-lang/crystal#10239]
```crystal
a = [1,2,3]
iter = a.each

item = iter.next

case item

when 0..9
    puts "#{item}: digit"
    item = iter.next
when Iterator::Stop
    puts "End Iteration"
end
```
in when 0..9, I think item is a integer but this code produce Error undefined method '>=' for Iterator::Stop (compile-time type is (Int32 | Iterator::Stop).
Put the when Iterator::Stop first, It compiles with no errors.

In (0..9).includes?(item), though item can't compare with each range value, how about Range returning False?

**Comment by straight-shoota** [comment]
Indeed. The default implementation of the case equality operator in `Object#===` is not strict about types (it delegates to `Object#==`) and so are most other implementations.
The trouble with `Range#===`/`Range#includes?` is that they're based on comparison operator between range begin/end and the method argument. But it's not trivial to restrict the method to types for which that operator is defined. It can't be restricted to the ranges generic arguments, because `(0..9).includes?(1.0)` should definitely work. A restriction to `Comparable` with appropriate type arguments might work.
#2549 would make this really easy.

**Comment by asterite** [comment]
We could make `when 0..9` be rewritten to `item.is_a?(Range) && (0..9) === item` because `0..9` can only mean a range. But maybe it's too much.

@nemorami If you put the `when Iterator::Stop` branch first then it works fine because the other branch won't have that type.

**Comment by straight-shoota** [comment]
But `item` isn't a `Range`. It's `Int32`. And we don't need to rewrite `when`, we just need `Range#===` to return `false` for types that it can't compare with begin and end.

**Comment by asterite** [comment]
Oh, yeah, my bad. I don't know what I was thinking...

**Comment by asterite** [comment]
I don't know how this can be done. `<=>` isn't defined for all types (I think it should, and the default implement in `Object` should return `nil`).

**Comment by straight-shoota** [comment]
I would strictly prefer #2549 because it's more versatile. It should be possible to avoid comparisons between types that don't possibly compare.

**Comment by asterite** [comment]
@straight-shoota Okay, but there's no consensus on #2549 . What would the syntax to do it be, and how can we keep it backwards-compatible?

**Comment by straight-shoota** [comment]
I think there's a strong consens to have this feature. The details just need to be figured out and agreed on.

**Comment by straight-shoota** [comment]
As I had feared, using `Comparable` to restrict the type isn't really feasible at the moment. Number primitives don't include `Comparable`. Maybe they should, though?
